gravitee-gateway-policy/src/main/java/io/gravitee/gateway/policy/impl/ExecutablePolicy.java

```diff
@@ -31,13 +37,62 @@
 
     private final String id;
     private final Object policy;
-    private final Method headMethod, streamMethod;
+    private final MethodHandle headMethodHandle;
+    private final MethodHandle streamMethodHandle;
```

Comment:
@jhaeyaert Can you provide a JMH Benchmark to confirm that relying on MethodHandle is better than reflection?
According to some posts, it is more performant only if you are relying to static MethodHandles.

Reply:
I've made a test I'll push.

**In our case**, use of `MethodHandle` is definitely better than use of `Reflection`, especially because (I guess) we are calculating and reordering the argument once instead of doing it on each call.

However, using `MethodHandle` is obviously less performant than doing a direct call but seems still more performant than direct use of `Reflection` (means invoke method without passing by `ExecutionPolicy` with the cost of reordering the arguments).

**Synthesis**

```
Benchmark                                             Mode  Cnt        Score       Error   Units
ExecutablePolicyBenchmark.benchDirect                thrpt    5  1309915,061 ± 13197,391  ops/ms
ExecutablePolicyBenchmark.benchMethodHandle          thrpt    5   189506,610 ±  2029,553  ops/ms
ExecutablePolicyBenchmark.benchReflection            thrpt    5     7835,654 ±    49,205  ops/ms
ExecutablePolicyBenchmark.benchReflectionDirectCall  thrpt    5   118097,933 ±  2598,028  ops/ms
```

**Detailed results**

Below you will find the different results.

```
# JMH version: 1.26
# VM version: JDK 11.0.5, Java HotSpot(TM) 64-Bit Server VM, 11.0.5+10-LTS
# VM invoker: /Library/Java/JavaVirtualMachines/jdk-11.0.5.jdk/Contents/Home/bin/java
# VM options: -javaagent:/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=63445:/Applications/IntelliJ IDEA.app/Contents/bin -Dfile.encoding=UTF-8
# Warmup: 5 iterations, 10 s each
# Measurement: 5 iterations, 10 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Throughput, ops/time
# Benchmark: io.gravitee.gateway.policy.impl.ExecutablePolicyBenchmark.benchDirect

# Run progress: 0,00% complete, ETA 00:06:40
# Fork: 1 of 1
# Warmup Iteration   1: 1269962,861 ops/ms
# Warmup Iteration   2: 1292648,056 ops/ms
# Warmup Iteration   3: 1300211,052 ops/ms
# Warmup Iteration   4: 1305300,374 ops/ms
# Warmup Iteration   5: 1306944,112 ops/ms
Iteration   1: 1309292,832 ops/ms
Iteration   2: 1313965,707 ops/ms
Iteration   3: 1312956,999 ops/ms
Iteration   4: 1306629,540 ops/ms
Iteration   5: 1306730,227 ops/ms


Result "io.gravitee.gateway.policy.impl.ExecutablePolicyBenchmark.benchDirect":
  1309915,061 ±(99.9%) 13197,391 ops/ms [Average]
  (min, avg, max) = (1306629,540, 1309915,061, 1313965,707), stdev = 3427,320
  CI (99.9%): [1296717,670, 1323112,452] (assumes normal distribution)


# JMH version: 1.26
# VM version: JDK 11.0.5, Java HotSpot(TM) 64-Bit Server VM, 11.0.5+10-LTS
# VM invoker: /Library/Java/JavaVirtualMachines/jdk-11.0.5.jdk/Contents/Home/bin/java
# VM options: -javaagent:/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=63445:/Applications/IntelliJ IDEA.app/Contents/bin -Dfile.encoding=UTF-8
# Warmup: 5 iterations, 10 s each
# Measurement: 5 iterations, 10 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Throughput, ops/time
# Benchmark: io.gravitee.gateway.policy.impl.ExecutablePolicyBenchmark.benchMethodHandle

# Run progress: 25,00% complete, ETA 00:05:03
# Fork: 1 of 1
# Warmup Iteration   1: 185145,568 ops/ms
# Warmup Iteration   2: 186714,208 ops/ms
# Warmup Iteration   3: 190477,421 ops/ms
# Warmup Iteration   4: 190051,896 ops/ms
# Warmup Iteration   5: 189394,148 ops/ms
Iteration   1: 189924,229 ops/ms
Iteration   2: 190021,664 ops/ms
Iteration   3: 189459,145 ops/ms
Iteration   4: 188691,535 ops/ms
Iteration   5: 189436,478 ops/ms


Result "io.gravitee.gateway.policy.impl.ExecutablePolicyBenchmark.benchMethodHandle":
  189506,610 ±(99.9%) 2029,553 ops/ms [Average]
  (min, avg, max) = (188691,535, 189506,610, 190021,664), stdev = 527,068
  CI (99.9%): [187477,057, 191536,163] (assumes normal distribution)


# JMH version: 1.26
# VM version: JDK 11.0.5, Java HotSpot(TM) 64-Bit Server VM, 11.0.5+10-LTS
# VM invoker: /Library/Java/JavaVirtualMachines/jdk-11.0.5.jdk/Contents/Home/bin/java
# VM options: -javaagent:/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=63445:/Applications/IntelliJ IDEA.app/Contents/bin -Dfile.encoding=UTF-8
# Warmup: 5 iterations, 10 s each
# Measurement: 5 iterations, 10 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Throughput, ops/time
# Benchmark: io.gravitee.gateway.policy.impl.ExecutablePolicyBenchmark.benchReflection

# Run progress: 50,00% complete, ETA 00:03:22
# Fork: 1 of 1
# Warmup Iteration   1: 7449,416 ops/ms
# Warmup Iteration   2: 7877,168 ops/ms
# Warmup Iteration   3: 7834,675 ops/ms
# Warmup Iteration   4: 7819,158 ops/ms
# Warmup Iteration   5: 7777,096 ops/ms
Iteration   1: 7820,386 ops/ms
Iteration   2: 7833,371 ops/ms
Iteration   3: 7851,386 ops/ms
Iteration   4: 7845,665 ops/ms
Iteration   5: 7827,462 ops/ms


Result "io.gravitee.gateway.policy.impl.ExecutablePolicyBenchmark.benchReflection":
  7835,654 ±(99.9%) 49,205 ops/ms [Average]
  (min, avg, max) = (7820,386, 7835,654, 7851,386), stdev = 12,778
  CI (99.9%): [7786,449, 7884,859] (assumes normal distribution)


# JMH version: 1.26
# VM version: JDK 11.0.5, Java HotSpot(TM) 64-Bit Server VM, 11.0.5+10-LTS
# VM invoker: /Library/Java/JavaVirtualMachines/jdk-11.0.5.jdk/Contents/Home/bin/java
# VM options: -javaagent:/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=63445:/Applications/IntelliJ IDEA.app/Contents/bin -Dfile.encoding=UTF-8
# Warmup: 5 iterations, 10 s each
# Measurement: 5 iterations, 10 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Throughput, ops/time
# Benchmark: io.gravitee.gateway.policy.impl.ExecutablePolicyBenchmark.benchReflectionDirectCall

# Run progress: 75,00% complete, ETA 00:01:41
# Fork: 1 of 1
# Warmup Iteration   1: 259794,020 ops/ms
# Warmup Iteration   2: 106804,920 ops/ms
# Warmup Iteration   3: 119678,959 ops/ms
# Warmup Iteration   4: 119603,930 ops/ms
# Warmup Iteration   5: 117179,872 ops/ms
Iteration   1: 117293,822 ops/ms
Iteration   2: 117901,618 ops/ms
Iteration   3: 119060,114 ops/ms
Iteration   4: 118440,489 ops/ms
Iteration   5: 117793,620 ops/ms


Result "io.gravitee.gateway.policy.impl.ExecutablePolicyBenchmark.benchReflectionDirectCall":
  118097,933 ±(99.9%) 2598,028 ops/ms [Average]
  (min, avg, max) = (117293,822, 118097,933, 119060,114), stdev = 674,700
  CI (99.9%): [115499,905, 120695,960] (assumes normal distribution)


# Run complete. Total time: 00:06:44

REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
experiments, perform baseline and negative tests that provide experimental control, make sure
the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
Do not assume the numbers tell you what you want them to tell.

Benchmark                                             Mode  Cnt        Score       Error   Units
ExecutablePolicyBenchmark.benchDirect                thrpt    5  1309915,061 ± 13197,391  ops/ms
ExecutablePolicyBenchmark.benchMethodHandle          thrpt    5   189506,610 ±  2029,553  ops/ms
ExecutablePolicyBenchmark.benchReflection            thrpt    5     7835,654 ±    49,205  ops/ms
ExecutablePolicyBenchmark.benchReflectionDirectCall  thrpt    5   118097,933 ±  2598,028  ops/ms
```